### PR TITLE
feat: external completions for commands/flags

### PR DESCRIPTION
**Pull request by herlon214**
# Description

This PR adds support to use external completions (like [carapace](https://github.com/rsteube/carapace)) instead of Nushell completions by adding a new config called `external_completer` which is optional.
```
let carapace_completer = {|spans| 
    carapace $spans.0 nushell $spans | from json
}

# The default config record. This is where much of your global configuration is setup.
let-env config = {
  external_completer: $carapace_completer
```

Basically, if you set one `external_completer` it will use it instead of our current Nushell completion flow for **commands and flags.**

**Current issues:**
- [x] Flags aren't being handled (e.g: `git config --<tab>`)
- [x] Resolve alias before sending to external completer
- [x] Remove extra `a` added in the spans
- [ ] Open quotes aren't being handled (e.g: `chown "<TAB>`) <--- We will leave this one up to the `external_completer` Nu function handle

# Tests

Make sure you've done the following:

- [x] Add tests that cover your changes, either in the command examples, the crate/tests folder, or in the /tests folder.
- [x] Try to think about corner cases and various ways how your changes could break. Cover them with tests.
- [x] If adding tests is not possible, please document in the PR body a minimal example with steps on how to reproduce so one can verify your change works.

Make sure you've run and fixed any issues with these commands:

- [x] `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- [x] `cargo clippy --workspace --features=extra -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- [x] `cargo test --workspace --features=extra` to check that all the tests pass

### CODE REVIEW

File: crates/nu-cli/src/completions/completer.rs [hunk 1]
```diff
@@ -56,13 +57,93 @@ impl NuCompleter {
...
+            Ok(pd) => {
+                let value = pd.into_value(span);
+                match &value {
+                    Value::List { vals, span: _ } => {
+                        return map_value_completions(vals.iter(), span, offset)
+                    }
+                    _ => {}
+                }
+            }
+            Err(err) => println!("failed to eval completer block: {}", err),
```
Comment (herlon214) [inline crates/nu-cli/src/completions/completer.rs:127]: What should we do if we fail to eval the block?

File: crates/nu-cli/src/completions/completer.rs [hunk 2]
```diff
@@ -383,3 +464,65 @@ fn most_left_variable(
 
     Some((var, sublevels))
 }
+
+pub fn map_value_completions<'a>(
```
Comment (herlon214) [inline crates/nu-cli/src/completions/completer.rs:468]: I moved this fn from `Custom Completions` in order to be able to reuse it.

**Comment by herlon214** [comment]
Would also be nice to move the `$external_completer` to outside config, somehow.

File: crates/nu-cli/src/completions/completer.rs [hunk 3]
```diff
@@ -56,13 +57,90 @@ impl NuCompleter {
...
+            end: pos,
+        };
+
+        let line_pos = pos - offset;
+        let stack = self.stack.clone();
+
+        let block = self.engine_state.get_block(decl_id);
+        let mut callee_stack = stack.gather_captures(&block.captures);
+
+        for (idx, PositionalArg { var_id, .. }) in
```
Comment (elferherrera) [inline crates/nu-cli/src/completions/completer.rs:78]: Instead of using a for loop, why don't you use get?

File: crates/nu-cli/src/completions/completer.rs (hunk 1 above)
Comment (elferherrera) [reply crates/nu-cli/src/completions/completer.rs:124]: I'm of the idea that it should return a Result, but that will change more things in the code

File: crates/nu-cli/src/completions/completer.rs [hunk 4]
```diff
@@ -56,13 +57,90 @@ impl NuCompleter {
         suggestions
     }
 
+    fn external_completion(
+        &self,
+        decl_id: usize,
```
Comment (kubouch) [inline crates/nu-cli/src/completions/completer.rs:62]: This should be called `block_id: BlockId` for clarity.

File: crates/nu-cli/src/completions/completer.rs (hunk 3 above)
Comment (herlon214) [reply crates/nu-cli/src/completions/completer.rs]: true

File: crates/nu-cli/src/completions/completer.rs (hunk 1 above)
Comment (herlon214) [reply crates/nu-cli/src/completions/completer.rs:120]: Yeah, but then it goes back to the `fn completion_helper(&mut self, line: &str, pos: usize) -> Vec<Suggestion> {` which should handle it anyway and return a `Vec<Suggestion>`

File: crates/nu-cli/src/completions/completer.rs [hunk 5]
```diff
@@ -56,21 +57,95 @@ impl NuCompleter {
...
 
         for pipeline in output.pipelines.into_iter() {
             for expr in pipeline.expressions {
                 let flattened: Vec<_> = flatten_expression(&working_set, &expr);
                 let span_offset: usize = alias_offset.iter().sum();
+                let mut spans: Vec<String> = vec![];
 
                 for (flat_idx, flat) in flattened.iter().enumerate() {
+                    // Skip the last 'a' as span item
+                    if flat_idx != flattened.len() {
```
Comment (rsteube) [inline crates/nu-cli/src/completions/completer.rs:141]: Just tried the new version and this didn't work for me (still had the 'a' suffix).

This seems to work:
```rust
                    let current_span_str = String::from_utf8_lossy(&current_span);
                    // Skip the last 'a' as span item
                    if flat_idx == flattened.len() - 1 {
                        let mut chars = current_span_str.chars();
                        chars.next_back();
                        let current_span_str = chars.as_str().to_owned();
                        spans.push(current_span_str.to_string());
                    } else {
                        spans.push(current_span_str.to_string());
                    }

```

```sh
let external_completer = {|spans|
   carapace $spans.0 nushell $spans | from json
}
```

Comment (rsteube) [reply crates/nu-cli/src/completions/completer.rs:141]: Open quotes aren't handled though.
```sh
chown ""<TAB> # works ["chown", ""]
chown "<TAB>  # fails ["chown", "\""]
```

Comment (herlon214) [reply crates/nu-cli/src/completions/completer.rs:141]: Yeah for the `'a'` I forgot to add the ` -1`

Comment (herlon214) [reply crates/nu-cli/src/completions/completer.rs]: > Open quotes aren't handled though.
> 
> ```shell
> chown ""<TAB> # works ["chown", ""]
> chown "<TAB>  # fails ["chown", "\""]
> ```

I added the issue in the description so I won't forget

File: crates/nu-cli/src/completions/completer.rs [hunk 6]
```diff
@@ -56,21 +57,100 @@ impl NuCompleter {
...
+
+        vec![]
+    }
+
     fn completion_helper(&mut self, line: &str, pos: usize) -> Vec<Suggestion> {
         let mut working_set = StateWorkingSet::new(&self.engine_state);
         let offset = working_set.next_span_start();
         let (mut new_line, alias_offset) = try_find_alias(line.as_bytes(), &working_set);
         let initial_line = line.to_string();
         new_line.push(b'a');
```
Comment (kubouch) [inline crates/nu-cli/src/completions/completer.rs:126]: What is this "a" for?

**Comment by herlon214** [comment]
I tested with alias and looks like it's working:
```
$ alias l = ls
$ l src/
src/README.md        src/config_files.rs  src/logger.rs        src/main.rs
src/test_bins.rs     src/tests.rs         src/tests/
```

File: crates/nu-cli/src/completions/completer.rs (hunk 6 above)
Comment (kubouch) [reply crates/nu-cli/src/completions/completer.rs:126]: Ah, ok, I see it's not your change, it just surprised me.

File: crates/nu-cli/src/completions/completer.rs [hunk 7]
```diff
@@ -89,6 +169,11 @@ impl NuCompleter {
                             }
                         };
 
+                        // Check if external completer
```
[the remark this answers is not in the document]
Comment (herlon214) [reply crates/nu-cli/src/completions/completer.rs:172]: IMO if you set one external completer you don't wanna rely on Nushell completion at all

[the remark this answers is not in the document]
Comment (herlon214) [reply crates/nu-cli/src/completions/completer.rs:172]: It's a bit tricky to move it

[the remark this answers is not in the document]
Comment (herlon214) [reply crates/nu-cli/src/completions/completer.rs:172]: Perhaps we should move it to before (or after) command completion?

[the remark this answers is not in the document]
Comment (rsteube) [reply crates/nu-cli/src/completions/completer.rs:172]: The builtins are currently also being completed by the external one (e.g. nushells `ls`).

[the remark this answers is not in the document]
Comment (herlon214) [reply crates/nu-cli/src/completions/completer.rs:172]: > Perhaps we should move it to before (or after) command completion?

Hmm, I was testing here, it's actually tricky.
Wherever we put the external completer after (e.g: after handling file completions), that will unable the external completer do complete that.

[the remark this answers is not in the document]
Comment (kubouch) [reply crates/nu-cli/src/completions/completer.rs:172]: I think, actually, it should be limited to only complete external commands. We shouldn't use external completers for completing Nushell's internals like built-in commands, variables, etc.

[the remark this answers is not in the document]
Comment (herlon214) [reply crates/nu-cli/src/completions/completer.rs]: Okay, added only for external commands then

[the remark this answers is not in the document]
Comment (herlon214) [reply crates/nu-cli/src/completions/completer.rs]: * and flags

[the remark this answers is not in the document]
Comment (rsteube) [reply crates/nu-cli/src/completions/completer.rs]: Path/Command completion seems to be broken:
```sh
<TAB>
../<TAB>
unknowncmd /<TAB> 
```

**Comment by kubouch** [comment]
I'll download this PR and look into it carefully hopefully tomorrow. I feel like we're almost there, just need to find the right spot for it to not interfere with Nushell's own completions.

File: crates/nu-cli/src/completions/completer.rs [hunk 8]
```diff
@@ -56,21 +57,100 @@ impl NuCompleter {
         suggestions
     }
 
+    fn external_completion(
+        &self,
+        block_id: usize,
```
Comment (kubouch) [inline crates/nu-cli/src/completions/completer.rs:62]: The type should be BlockId to make it clearer.

File: crates/nu-cli/src/completions/completer.rs [hunk 9]
```diff
@@ -113,6 +193,11 @@ impl NuCompleter {
 
                         // Flags completion
                         if prefix.starts_with(b"-") {
+                            // Check if external completer
+                            if let Some(decl_id) = config.external_completer {
```
Comment (kubouch) [inline crates/nu-cli/src/completions/completer.rs:197]: The `decl_id`s should be called `block_id`s. It's a bit pedantic but it might be confusing if someone reads the code later.

File: crates/nu-cli/src/completions/completer.rs [hunk 10]
```diff
@@ -206,6 +291,12 @@ impl NuCompleter {
                                 );
                             }
                             flat_shape => {
+                                // Check if external completer
+                                if let Some(decl_id) = config.external_completer {
```
Comment (kubouch) [inline crates/nu-cli/src/completions/completer.rs:295]: same, `decl_id` should be `block_id`

File: crates/nu-cli/src/completions/completer.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (herlon214) [reply crates/nu-cli/src/completions/completer.rs]: nice catch @rsteube 
using `cd ../<TAB>` works tho 🤔

[the remark this answers is not in the document]
Comment (herlon214) [reply crates/nu-cli/src/completions/completer.rs]:
```rust
flat_shape => {
                                // Check if external completer
                                if let Some(decl_id) = config.external_completer {
                                    return self
                                        .external_completion(decl_id, spans, offset, new_span);
                                }

                                let mut completer = CommandCompletion::new(
                                    self.engine_state.clone(),
                                    &working_set,
                                    flattened.clone(),
                                    // flat_idx,
                                    flat_shape.clone(),
                                );

                                let out: Vec<_> = self.process_completion(
                                    &mut completer,
                                    &working_set,
                                    prefix.clone(),
                                    new_span,
                                    offset,
                                    pos,
                                );

                                if out.is_empty() {
                                    let mut completer =
                                        FileCompletion::new(self.engine_state.clone());

                                    return self.process_completion(
                                        &mut completer,
                                        &working_set,
                                        prefix,
                                        new_span,
                                        offset,
                                        pos,
                                    );
                                }

                                return out;
                            }
```

I think that part is done inside the `if out.is_empty() {`

**Comment by kubouch** [comment]
I found these issues testing this PR:
1. `ls -<tab>` will trigger the external completer.
2. Just `<tab>` on an empty line, or `l<tab>` also triggers the external completion.

In both cases, the external completer should be triggered after trying Nushell's completions, not before, because it prevents us from completing built-in commands and flags correctly.

**Comment by herlon214** [comment]
@kubouch what do you mean by moving them lower in priority?
If we trigger Nushell ones it automatically returns that result and will never reach the external one.

We need to check if nushell internals return empty, if so we trigger the external, is that what you mean?

**Comment by kubouch** [comment]
> @kubouch what do you mean by moving them lower in priority? If we trigger Nushell ones it automatically returns that result and will never reach the external one.
> 
> We need to check if nushell internals return empty, if so we trigger the external, is that what you mean?

Yeah, that's what I mean. Check if Nushell completions return something and if not, try the external completer.

**Comment by herlon214** [comment]
> > @kubouch what do you mean by moving them lower in priority? If we trigger Nushell ones it automatically returns that result and will never reach the external one.
> > We need to check if nushell internals return empty, if so we trigger the external, is that what you mean?
> 
> Yeah, that's what I mean. Check if Nushell completions return something and if not, try the external completer.

Got it, I will update the PR then

**Comment by herlon214** [comment]
I found one issue with the current code, once we don't have completions for `gh alias` nu should trigger the external compelter. But instead it's returning a list of all commands:
Screen Shot 2022-08-19 at 1 16 01 PM

**Comment by rsteube** [comment]
There also seems to be a fundamental step missing compared to how nushell passed arguments to a command.
I think the slicing is correct but there should be some way to evalute the slices.

Here `'good first issue',` is passed with the quotes:
```sh
gh issue list --label 'good first issue',<TAB>
# ''good first issue',0.67'                     Issues that should be fixed in v0.67
# ''good first issue',Stale'                    used for marking issues and prs as stale
```

More apparent with the string interpolation which results slices kinda like this:
```sh
ls $"($env.HOME)"
# [ 'ls', '$"', '(', '$env', 'HOME', ')', '"a' ]
```

**Comment by kubouch** [comment]
@rsteube The eval step for the argument is a good idea, but I would leave it for a future work. I'd rather get this PR merged first for simple cases.

About the quotes, ` 'good first issue',` is one string from Nushell's point of view. It doesn't know `,` is supposed to split the arguments, that's just an interpretation of the external command that we can't assume. We might want to expand the external command parsing to be smarter, but we'd have to change Nushell parsing rules which is also outside the scope of this PR. Ideally, there should be no parsing involved in the completions code, we should just reuse existing logic, otherwise, we'd create a mess.

**Comment by kubouch** [comment]
> I found one issue with the current code, once we don't have completions for `gh alias` nu should trigger the external compelter. But instead it's returning a list of all commands: Screen Shot 2022-08-19 at 1 16 01 PM

I think the error here might be related to the long-standing bug that if there is an empty input, it defaults to completing all the commands. Like when you press `<tab>` on an empty line, you get a screen full of commands which is not very useful. Also, completing external commands when you already typed some words (like `gh alias`) is almost always not what you want.

You want to complete commands only in two cases:
1. After you typed more than 0 non-space letters 
2. All the words before the cursor position match a known command or subcommand (like completing `path exp<tab>`, you should get `path expand` suggestion). If there is no match for a known command / subcommand, then no command completion should be given and external completion would kick in.

Also, it seems to me, the file completion should have a priority over command completion. Currently, it first tries commands, then files but I'm wondering if it should be the other way.

So, to sum up, it seems the error here is in the completion logic itself, not in introducing the external completer. What we could do is to land this PR and make a new PR to fix the other completion issues. However, we should probably make some list to document the cases where we know the completions are broken.

There is also https://github.com/nushell/nushell/pull/6257 which might be about fixing the same command completion issue. It introduces a new syntax shape, though, which we thought might not be the best approach. You might want to take a look at it and maybe work together with @merelymyself if you want to address the wrong command completions.

**Comment by herlon214** [comment]
How can we have a minimum working version so the users can test and give some feedback. Maybe we can address some corner cases later given user's feedback?

**Comment by kubouch** [comment]
Yes, the `gh alias` issue is already present without your modification, so we can leave it for later. There is still some failing test in this PR, though.

**Comment by herlon214** [comment]
Yeah @kubouch I think that test failing is related to the issue we discussed. Should I remove that test then?

**Comment by kubouch** [comment]
OK, if it's related to the `gh alias` issue, then let's `#[ignore]` the test for now. Could you please create an issue that documents the problem.

I think if you prioritize the external completer, it would call to it before completing the Nushell command which is not what we want. I suspect prioritizing the external completer would need a bit of extra logic I outlined earlier to make it work reliably. Like `path exp<tab>` should auto-complete the Nushell command first and not call external completer at all.

And one more thing: Because you added a new config option, could you please put an example entry to [the default config](https://github.com/nushell/nushell/blob/main/crates/nu-utils/src/sample_config/default_config.nu) (commented out or set to `$nothing` by default)? It might be worth adding to the book somewhere as well once it gets merged.

**Comment by kubouch** [comment]
OK, let's give this a try! Thanks for the effort!

**Comment by rsteube** [comment]
@herlon214 The merged version doesn't work for me (`main` branch). Did something change?

**Comment by herlon214** [comment]
@rsteube I think the priority of the completions changed, and if I'm not mistaken currently it isn't working well due to one issue being fixed on https://github.com/nushell/nushell/pull/6257
